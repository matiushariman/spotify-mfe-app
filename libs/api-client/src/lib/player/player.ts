import axios, { AxiosInstance, AxiosError, isAxiosError } from 'axios';

import { API_BASE_URL } from '../../constants';

import type {
  GetRecentlyPlayedTracksResponse,
  GetRecentlyPlayedTracksError,
} from './player.types';
import type { ApiClassConstructor } from '../../types/ApiClassConstructor.types';

export class PlayerApi {
  private readonly http: AxiosInstance;

  constructor({ token }: ApiClassConstructor) {
    this.http = axios.create({
      baseURL: `${API_BASE_URL}/v1`,
      headers: {
        'Content-Type': 'application /json',
        Authorization: `Bearer ${token}`,
      },
    });
  }

  /**
   * Get tracks from the current user's recently played tracks.
   *
   * - {@link https://developer.spotify.com/documentation/web-api/reference/#/operations/get-recently-played|API Reference - Get Recently Played Tracks}
   */
  async getRecentlyPlayedTracks() {
    return this.http
      .get<GetRecentlyPlayedTracksResponse>('/me/player/recently-played')
      .then((res) => res)
      .catch((err: Error | AxiosError<GetRecentlyPlayedTracksError>) => {
        if (isAxiosError(err)) {
          return Promise.reject(err?.response?.data);
        }

        return Promise.reject(err);
      });
  }
}
