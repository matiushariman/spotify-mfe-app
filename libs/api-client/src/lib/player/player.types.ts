import { ApiErrorResponse } from '../../types/ApiErrorResponse.types';

export interface GetRecentlyPlayedTracksResponse {
  /**
   * - A link to the Web API endpoint returning the full result of the request.
   */
  readonly href: string;
  /**
   * - The requested data.
   */
  readonly items: string[];
  /**
   * - The maximum number of items in the response (as set in the query or by default).
   */
  readonly limit: number;
  /**
   * - URL to the next page of items. ( `null` if none)
   */
  readonly next: string | null;
  /**
   * - The cursors used to find the next set of items.
   */
  readonly cursors: {
    /**
     * - The cursor to use as key to find the next page of items.
     */
    readonly after: string;
  };
  /**
   * - The total number of items available to return.
   */
  readonly total: number;
}

export type GetRecentlyPlayedTracksError = ApiErrorResponse;
