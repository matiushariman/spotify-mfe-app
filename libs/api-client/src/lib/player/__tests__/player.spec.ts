import { PlayerApi } from '../player';

describe('lib/player', () => {
  const playerApi = new PlayerApi({ token: '' });

  describe('getRecentlyPlayedTracks()', () => {
    it('should return recenty played tracks data', async () => {
      const res = await playerApi.getRecentlyPlayedTracks();

      expect(res.status).toEqual(200);
      expect(res.data).toMatchSnapshot();
    });
  });
});
