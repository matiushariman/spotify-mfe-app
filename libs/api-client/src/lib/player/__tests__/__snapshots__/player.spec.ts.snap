// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/player getRecentlyPlayedTracks() should return recenty played tracks data 1`] = `
Object {
  "cursors": Object {
    "after": "string",
  },
  "href": "string",
  "items": Array [
    "",
  ],
  "limit": 0,
  "next": "string",
  "total": 0,
}
`;
